### PR TITLE
test update

### LINKED ISSUE #22: Tests
All of the new features will need some amount of tests added for show in the project turn in.

- Phishing protection #10 
- Stats page updates #13 #15 
- Location and browser insights #14 
- Map view #20 

Refer to closed pull requests above to see what new code needs to be covered.

### PR DESCRIPTION
**Pull request by hunglequoc2001**
I added test for #10 and #14, not all worked yet but function name should cover  most of test cases, I will commit a fixed version soon

### CODE REVIEW

**Comment by hunglequoc2001** [comment]
Test for #13  seems to already been covered, however, I'm not so sure how to create test cases for #15 and #20

File: url_shortner_server/shortner/tests/test_security.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        self.assertEqual(self.link.vt_analysis_stats, {"malicious": 0, "suspicious": 0})
+
+    @patch("shortner.utils.Client")
+    def test_virus_total_api_error(self, mock_client):
+        """Test that API errors in VirusTotal don't crash the function."""
+        mock_client.side_effect = Exception("VirusTotal API error")
+
+        do_security_check(self.link)
+
+        self.assertFalse(self.link.possibly_malicious)  # Should remain unaffected
```
Comment (bradenhelmer) [inline url_shortner_server/shortner/tests/test_security.py:73]: The mocking logic is incorrect here, the vt.Client should be mocked and not shortener.utils.Client.

File: url_shortner_server/shortner/tests/test_location.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
+import unittest
+from unittest.mock import patch, MagicMock
+from shortner.stub_view import StubView
```
Comment (bradenhelmer) [inline url_shortner_server/shortner/tests/test_location.py:3]: should be `from shortner.views import StubView`

File: url_shortner_server/shortner/tests/test_location.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        result = StubView.get_location(self.valid_ip)
+
+        # Assertions
+        self.assertEqual(result, {})
+        mock_get.assert_called_once_with(
+            f"https://ipinfo.io/{self.valid_ip}/json?token={self.mock_token}", timeout=30
+        )
+
+    @patch("shortner.views.stub_view.getenv")
+    @patch("shortner.views.stub_view.requests.get")
```
Comment (bradenhelmer) [inline url_shortner_server/shortner/tests/test_location.py:84]: These patches cannot be found, I believe you are referencing packages that don't exist.

**Pull request merged**